Mathlib/GroupTheory/Perm/Cycle/Basic.lean

```diff
@@ -1978,3 +2084,210 @@ theorem _root_.Finset.sum_mul_sum_eq_sum_perm (hσ : σ.IsCycleOn s) (f g : ι 
+          (g.subtypePerm (mem_cycleFactorsFinset_support hc)) := by
+  rw [IsCycle.commute_iff' (mem_cycleFactorsFinset_iff.mp hc).1]
+  apply exists_congr
+  intro hc'
+  simp only [Subgroup.mem_zpowers_iff]
+  apply exists_congr
+  intro n
+  unfold subtypePermOfSupport
+  rw [Equiv.Perm.subtypePerm_on_cycleFactorsFinset hc]
+  rfl
+
+theorem zpow_eq_ofSubtype_subtypePerm_iff
+    {g c : Equiv.Perm α} {s : Finset α}
+    (hg : ∀ x, x ∈ s ↔ g x ∈ s) (hc : c.support ⊆ s) (n : ℤ) :
+    c ^ n = ofSubtype (g.subtypePerm hg) ↔
+      c.subtypePerm (isInvariant_of_support_le hc) ^ n = g.subtypePerm hg := by
+  constructor
+  · intro h; ext ⟨x, hx⟩; let h' := Perm.congr_fun h x
+    simp only [h', subtypePerm_apply_zpow_of_mem, Subtype.coe_mk, subtypePerm_apply]
+    rw [ofSubtype_apply_of_mem]
+    simp only [Subtype.coe_mk, subtypePerm_apply]
+    exact hx
+  · intro h; ext x
+    rw [← h]
+    by_cases hx : x ∈ s
+    · rw [ofSubtype_apply_of_mem]
+      simp only [subtypePerm_zpow, subtypePerm_apply]
+      exact hx
+    · rw [ofSubtype_apply_of_not_mem, ← not_mem_support]
+      exact fun hx' ↦ hx (hc (support_zpow_le _ _ hx'))
+      exact hx
+
+theorem cycle_zpow_mem_support_iff {g : Perm α}
+    (hg : g.IsCycle) {n : ℤ} {x : α} (hx : g x ≠ x) :
+    (g ^ n) x = x ↔ n % g.support.card = 0 := by
+  let q := n / g.support.card; let r := n % g.support.card
+  change _ ↔ r = 0
+  have div_euc : r + g.support.card * q = n ∧ 0 ≤ r ∧ r < g.support.card := by
+    rw [← Int.ediv_emod_unique _]
+    constructor; rfl; rfl
+    simp only [Int.coe_nat_pos]
+    apply lt_of_lt_of_le _ (IsCycle.two_le_card_support hg); norm_num
+  simp only [← hg.orderOf] at div_euc
+  obtain ⟨m, hm⟩ := Int.eq_ofNat_of_zero_le div_euc.2.1
+  simp only [hm, Nat.cast_nonneg, Nat.cast_lt, true_and_iff] at div_euc
+  simp only [hm, Nat.cast_eq_zero]
+  rw [← div_euc.1, zpow_add g, zpow_mul, zpow_ofNat]
+  simp only [pow_orderOf_eq_one, zpow_ofNat, one_zpow, mul_one]
+  have : (g ^ m) x = x ↔ g ^ m = 1 := by
+    constructor
+    · intro hgm
+      simp [IsCycle.pow_eq_one_iff hg]
+      use x
+    · intro hgm; rw [hgm]; simp only [coe_one, id.def]
+  rw [this, div_euc.1, ← hg.orderOf, hm]
+  cases' dec_em (m = 0) with hm0 hm0'
+  · simp only [hm0, pow_zero, Nat.cast_zero]
+  · simp only [Nat.cast_eq_zero, hm0', iff_false]
+    exact pow_ne_one_of_lt_orderOf' hm0' div_euc.2
+
+theorem zpow_eq_zpow_on_iff (g : Perm α)
+    (m n : ℤ) (x : α) (hx : g x ≠ x) :
+    (g ^ m) x = (g ^ n) x ↔
+      m % (g.cycleOf x).support.card = n % (g.cycleOf x).support.card := by
+  rw [Int.emod_eq_emod_iff_emod_sub_eq_zero]
+  conv_lhs => rw [← Int.sub_add_cancel m n, Int.add_comm, zpow_add]
+  simp only [coe_mul, Function.comp_apply, EmbeddingLike.apply_eq_iff_eq]
+  rw [← Int.dvd_iff_emod_eq_zero]
+  rw [← cycleOf_zpow_apply_self g x, cycle_zpow_mem_support_iff]
+  · rw [← Int.dvd_iff_emod_eq_zero]
+  · exact isCycle_cycleOf g hx
+  · simp only [mem_support, cycleOf_apply_self]; exact hx
```

Comment:
I'll have a look and thing at it.

Reply:
@riccardobrasca had already asked for that (for another PR!). There are many sensible places to split this file, it does too many things at once. In particular, since it consists of 5 independent sections, each of them could constitute one file. 
And Basic.lean could just import all of them so that the rest is unchanged.